### PR TITLE
Add features to cover different usecases

### PR DESCRIPTION
**Pull request by Nemo157**
## Description
See the updated docs in lib.rs for details on the design.

## Motivation and Context
This allows users to only pay compilation cost for the compressions and usecases they care about (e.g. Tide can now stop compiling the `bufread` implementations it doesn't use).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the [CONTRIBUTING](https://github.com/rust-net-web/tide/blob/master/.github/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -11,14 +11,24 @@ description = """
...
-bytes = "0.4.12"
-flate2 = "1.0.7"
+brotli2 = { version = "0.3.2", optional = true }
+bytes = { version = "0.4.12", optional = true }
+flate2 = { version = "1.0.7", optional = true }
 futures-preview = "0.3.0-alpha.16"
 pin-project = "0.3.2"
-zstd = "0.4"
-zstd-safe = "1.4"
+libzstd = { version = "0.4", optional = true, package = "zstd", default-features = false }
```
Comment (Nemo157) [inline Cargo.toml]: The renaming of the package here is a little bit of a hack to allow a feature with the same name as an optional dependency.

Comment (fairingrey) [reply Cargo.toml]: Fair enough, I don't mind the disambiguation here.

**Comment by fairingrey** [comment]
@Nemo157 Travis freaks out over a `cargo fmt --all -- --check`, want to just run that real quick? I can also commit the change to this PR, I believe, but just checking with you.

File: src/stream/zstd.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@ use std::{
 use bytes::{Bytes, BytesMut};
 use futures::{ready, stream::Stream};
 use pin_project::unsafe_project;
```
Comment (fairingrey) [inline src/stream/zstd.rs]: Travis freaks out on this line, `use pin_project::unsafe_project;` should be below libzstd

**Comment by fairingrey** [comment]
Other than that, it's nice they're all included by default but can be reduced to minimal dependencies if a user calls for it. I'm a fan of the change.

**Comment by Nemo157** [comment]
@fairingrey thanks for fixing that up! I should probably have run the tests before actually checking that box...

**Pull request merged**